File: scratch/build_catalog_reconciliation.mjs
import fs from 'node:fs/promises';
import { SpreadsheetFile, Workbook } from '@oai/artifact-tool';

const root = process.cwd();
const rows = JSON.parse(await fs.readFile(`${root}/scratch/catalog_reconciliation.json`, 'utf8'));
const candidates = JSON.parse(await fs.readFile(`${root}/scratch/new_product_candidates.json`, 'utf8'));
const outputDir = `${root}/outputs/catalog_reconciliation`;
await fs.mkdir(outputDir, { recursive: true });

const workbook = Workbook.create();
const summary = workbook.worksheets.add('Resumen');
const sheet = workbook.worksheets.add('Conciliación');
const candidatesSheet = workbook.worksheets.add('Nuevos candidatos');

summary.getRange('A1:D1').merge();
summary.getRange('A1').values = [['Conciliación oficial: Excel + fotografías']];
summary.getRange('A1:D1').format = {
  fill: '#1959AD', font: { bold: true, color: '#FFFFFF', size: 14 },
  horizontalAlignment: 'center', verticalAlignment: 'center',
};
summary.getRange('A3:B6').values = [
  ['Productos oficiales (Excel)', rows.length],
  ['Conservar código de fábrica', rows.filter(r => r.estado === 'CONSERVAR CÓDIGO DE FÁBRICA').length],
  ['Revisar coincidencia', rows.filter(r => r.estado === 'REVISAR COINCIDENCIA').length],
  ['Generar QR interno', rows.filter(r => r.estado === 'GENERAR QR INTERNO').length],
];
summary.getRange('A3:A6').format = { fill: '#EAF2FF', font: { bold: true } };
summary.getRange('A3:B6').format.borders = { preset: 'all', style: 'thin', color: '#D1D5DB' };
summary.getRange('A8:D10').merge();
summary.getRange('A8').values = [[
  'Regla: un código de fábrica se conserva únicamente cuando nombre/tipo y especificaciones no se contradicen. Las coincidencias incompletas quedan para revisión; si no hay evidencia se propone QR interno.'
]];
summary.getRange('A8:D10').format = { fill: '#FFF7E6', wrapText: true, verticalAlignment: 'center' };
summary.getRange('A:A').format.columnWidth = 34;
summary.getRange('B:B').format.columnWidth = 18;
summary.getRange('C:D').format.columnWidth = 18;
summary.getRange('A1:D1').format.rowHeight = 28;
summary.showGridLines = false;

const columns = [
  ['ITEM', 'item'], ['ALMACÉN', 'almacen'], ['TIPO', 'tipo_oficial'],
  ['PRODUCTO OFICIAL', 'producto_oficial'], ['UNIDAD', 'unidad'], ['STOCK EXCEL', 'stock_excel'],
  ['FOTO', 'foto_referencia'], ['ARCHIVO FOTO', 'archivo_foto'], ['CÓDIGO FÁBRICA PROPUESTO', 'codigo_fabrica_propuesto'],
  ['TIPO DE CÓDIGO', 'tipo_codigo'], ['MARCA', 'marca_propuesta'], ['MODELO', 'modelo_propuesto'],
  ['SUBTIPO', 'subtipo_propuesto'], ['ATRIBUTOS / DATOS FOTO', 'atributos_foto'],
  ['ESPECIFICACIONES COINCIDENTES', 'especificaciones_coincidentes'], ['CONFLICTO', 'conflicto_especificaciones'],
  ['CONFIANZA', 'confianza'], ['ESTADO', 'estado'], ['QR INTERNO PROPUESTO', 'qr_interno_propuesto'], ['NOTA', 'nota'],
];
sheet.getRangeByIndexes(0, 0, 1, columns.length).values = [columns.map(([name]) => name)];
sheet.getRangeByIndexes(1, 0, rows.length, columns.length).values = rows.map(row => columns.map(([, key]) => {
  const value = row[key] ?? '';
  return (key === 'codigo_fabrica_propuesto' && /^\d/.test(String(value))) ? `'${value}` : value;
}));
const used = sheet.getRangeByIndexes(0, 0, rows.length + 1, columns.length);
sheet.getRangeByIndexes(0, 0, 1, columns.length).format = {
  fill: '#1959AD', font: { bold: true, color: '#FFFFFF' }, wrapText: true,
  horizontalAlignment: 'center', verticalAlignment: 'center',
};
used.format.borders = { preset: 'insideHorizontal', style: 'thin', color: '#E5E7EB' };
sheet.getRange(`Q2:Q${rows.length + 1}`).format.numberFormat = '0.000';
sheet.getRange(`F2:F${rows.length + 1}`).format.numberFormat = '#,##0';
sheet.getRange(`R2:R${rows.length + 1}`).conditionalFormats.add('containsText', { text: 'CONSERVAR', format: { fill: '#DCFCE7', font: { color: '#166534', bold: true } } });
sheet.getRange(`R2:R${rows.length + 1}`).conditionalFormats.add('containsText', { text: 'REVISAR', format: { fill: '#FEF3C7', font: { color: '#92400E', bold: true } } });
sheet.getRange(`R2:R${rows.length + 1}`).conditionalFormats.add('containsText', { text: 'GENERAR', format: { fill: '#EAF2FF', font: { color: '#1D4ED8', bold: true } } });
sheet.freezePanes.freezeRows(1);
sheet.getRange(`A1:T${rows.length + 1}`).format.wrapText = true;
const widths = [8,16,13,42,10,12,8,34,26,15,20,26,26,52,24,24,11,28,28,42];
widths.forEach((width, index) => sheet.getRangeByIndexes(0, index, rows.length + 1, 1).format.columnWidth = width);
sheet.getRange(`A1:T1`).format.rowHeight = 38;
sheet.showGridLines = false;

const candidateColumns = [
  ['FOTO', 'foto'], ['PRODUCTO PROPUESTO', 'producto_propuesto'], ['CATEGORÍA', 'categoria'],
  ['MARCA', 'marca'], ['MODELO', 'modelo'], ['CÓDIGO DE FÁBRICA', 'codigo_fabrica'],
  ['DATOS TÉCNICOS', 'datos_tecnicos'], ['ESTADO', 'estado'], ['MOTIVO', 'motivo'],
];
candidatesSheet.getRangeByIndexes(0, 0, 1, candidateColumns.length).values = [candidateColumns.map(([name]) => name)];
candidatesSheet.getRangeByIndexes(1, 0, candidates.length, candidateColumns.length).values = candidates.map(row => candidateColumns.map(([, key]) => {
  const value = row[key] ?? '';
  return (key === 'codigo_fabrica' && /^\d/.test(String(value))) ? `'${value}` : value;
}));
candidatesSheet.getRange(`A1:I${candidates.length + 1}`).format.wrapText = true;
candidatesSheet.getRange('A1:I1').format = { fill: '#1959AD', font: { bold: true, color: '#FFFFFF' }, horizontalAlignment: 'center' };
candidatesSheet.getRange(`A1:I${candidates.length + 1}`).format.borders = { preset: 'insideHorizontal', style: 'thin', color: '#E5E7EB' };
candidatesSheet.getRange(`H2:H${candidates.length + 1}`).format = { fill: '#FEF3C7', font: { color: '#92400E', bold: true } };
[8, 28, 15, 20, 24, 24, 60, 26, 48].forEach((width, index) => candidatesSheet.getRangeByIndexes(0, index, candidates.length + 1, 1).format.columnWidth = width);
candidatesSheet.freezePanes.freezeRows(1);
candidatesSheet.showGridLines = false;

const check = await workbook.inspect({ kind: 'table', range: 'Conciliación!A1:T8', include: 'values', tableMaxRows: 8, tableMaxCols: 20 });
console.log(check.ndjson);
const preview = await workbook.render({ sheetName: 'Resumen', range: 'A1:D10', scale: 1.5, format: 'png' });
await fs.writeFile(`${outputDir}/preview.png`, new Uint8Array(await preview.arrayBuffer()));
const candidatesPreview = await workbook.render({ sheetName: 'Nuevos candidatos', range: 'A1:I12', scale: 1, format: 'png' });
await fs.writeFile(`${outputDir}/candidates_preview.png`, new Uint8Array(await candidatesPreview.arrayBuffer()));
const output = await SpreadsheetFile.exportXlsx(workbook);
await output.save(`${outputDir}/conciliacion_catalogo_actualizada.xlsx`);

File: scratch/build_catalog_reconciliation.py
import json
import re
import unicodedata
from difflib import SequenceMatcher
from pathlib import Path

import pandas as pd


ROOT = Path(__file__).resolve().parent.parent
EXCEL = ROOT / 'bd.xlsx'
PHOTOS = Path(
    r'C:\Users\MARI\AppData\Local\Temp\claude\C--Users-MARI-app-inventario-mobil'
    r'\3e6bd6c4-af40-4066-86ee-6878272007a4\scratchpad\catalogo.tsv'
)
OUTPUT = ROOT / 'scratch' / 'catalog_reconciliation.json'
NEW_CANDIDATES_OUTPUT = ROOT / 'scratch' / 'new_product_candidates.json'

# Fotos sin equivalente claro entre los 122 productos oficiales. Se proponen
# para alta, pero no se insertan automáticamente: el administrador confirma
# cada caso antes de incorporarlo al catálogo maestro.
NEW_CANDIDATE_PHOTOS = {
    3: 'Thinner acrílico: no existe ficha equivalente en el Excel.',
    13: 'Hidrolavadora portátil: no existe ficha equivalente en el Excel.',
    17: 'Caja de registro para pozo a tierra: no existe ficha equivalente.',
    35: 'Grasa/lubricante SKF: no existe ficha equivalente.',
    36: 'Espuma de poliuretano 750 ml: no existe ficha equivalente.',
    37: 'Espuma de poliuretano 340 ml: variante que requiere ficha propia.',
    40: 'Solvente dieléctrico: no existe ficha equivalente.',
    43: 'UPS Hikvision: no existe ficha UPS equivalente.',
    70: 'Filtro de ventilador: no existe ficha equivalente.',
    122: 'Crimpadora de terminales: distinta del alicate genérico.',
    125: 'Crimpadora RJ45: herramienta de red, distinta del alicate genérico.',
}


def normalized(value):
    value = unicodedata.normalize('NFKD', str(value)).encode('ascii', 'ignore').decode()
    value = value.upper()
    value = re.sub(r'[^A-Z0-9.]+', ' ', value)
    value = re.sub(r'(\d)([A-Z])', r'\1 \2', value)
    return re.sub(r'\s+', ' ', value).strip()


def meaningful_tokens(value):
    ignored = {
        'DE', 'LA', 'EL', 'Y', 'CON', 'PARA', 'EN', 'SIN', 'POR', 'UN', 'UNA',
        'UND', 'EQUIPO', 'FERRETERIA', 'EPP', 'SOLAR', 'TIPO', 'DEL', 'LOS', 'LAS',
        'CC', 'CA', 'DC', 'AC', 'MONEDA', 'UNIDAD', 'PRODUCTO',
    }
    return {token for token in normalized(value).split() if len(token) > 1 and token not in ignored}


def technical_specs(value):
    return set(re.findall(
        r'\b\d+(?:\.\d+)?\s*(?:KW|W|V|AH|A|MM|M|T|PSI|LTS|LT|DB|AWG|HP)\b',
        normalized(value),
    ))


def specs_by_unit(specifications, unit):
    return {value for value in specifications if value.endswith(f' {unit}')}


def is_support_photo(row):
    details = str(row['datos']).upper()
    return 'NO ES ITEM DE INVENTARIO' in details or 'NO ES PRODUCTO' in details


def canonical_photo_rows(photos):
    """Keep one evidence record per visible code, while retaining no-code photos."""
    result = []
    seen_codes = set()
    for _, photo in photos.sort_values('n').iterrows():
        if is_support_photo(photo):
            continue
        code = str(photo['codigo']).strip()
        if code and code != '-':
            if code in seen_codes:
                continue
            seen_codes.add(code)
        result.append(photo)
    return result


def score_match(product_name, photo):
    photo_text = ' '.join(
        str(photo.get(column, ''))
        for column in ('categoria', 'subtipo', 'marca', 'modelo', 'datos')
    )
    product_tokens = meaningful_tokens(product_name)
    photo_tokens = meaningful_tokens(photo_text)
    shared = product_tokens & photo_tokens
    token_score = len(shared) / max(1, len(product_tokens | photo_tokens))
    sequence_score = SequenceMatcher(None, normalized(product_name), normalized(photo_text)).ratio()
    product_specs = technical_specs(product_name)
    photo_specs = technical_specs(photo_text)
    shared_specs = product_specs & photo_specs
    spec_score = len(shared_specs) / max(1, len(product_specs))

    # Variador e inversor se usan indistintamente en la información histórica.
    subtype = normalized(photo.get('subtipo', ''))
    product_normalized = normalized(product_name)
    family_bonus = 0.0
    if ('INVERSOR' in product_normalized and ('INVERSOR' in subtype or 'VARIADOR' in subtype)):
        family_bonus = 0.12
    elif any(token in product_normalized for token in ('CABLE', 'GUANTE', 'CASCO', 'LENTE', 'PERNO', 'DISYUNTOR')) and any(
        token in subtype for token in ('CABLE', 'GUANTE', 'CASCO', 'LENTE', 'PERNO', 'DISYUNTOR')
    ):
        family_bonus = 0.12

    score = 0.35 * token_score + 0.20 * sequence_score + 0.35 * spec_score + family_bonus
    return score, sorted(shared_specs), len(shared)


def main():
    official = pd.read_excel(EXCEL, sheet_name='BASE DE DATOS ALMACEN')
    photos = pd.read_csv(PHOTOS, sep='\t').fillna('')
    evidence = canonical_photo_rows(photos)
    reconciliation = []

    for _, product in official.iterrows():
        ranked = []
        for photo in evidence:
            score, shared_specs, shared_tokens = score_match(product['PRODUCTO'], photo)
            ranked.append((score, shared_specs, shared_tokens, photo))
        ranked.sort(key=lambda item: item[0], reverse=True)
        score, shared_specs, shared_tokens, best = ranked[0]
        code = str(best['codigo']).strip()
        code_is_available = bool(code and code != '-')
        product_specs = technical_specs(product['PRODUCTO'])
        photo_specs = technical_specs(' '.join(
            str(best.get(column, ''))
            for column in ('subtipo', 'marca', 'modelo', 'datos')
        ))
        expected_voltage = specs_by_unit(product_specs, 'V')
        observed_voltage = specs_by_unit(photo_specs, 'V')
        expected_power = specs_by_unit(product_specs, 'KW') | specs_by_unit(product_specs, 'W')
        observed_power = specs_by_unit(photo_specs, 'KW') | specs_by_unit(photo_specs, 'W')
        has_conflicting_voltage = bool(
            expected_voltage and observed_voltage and not (expected_voltage & observed_voltage)
        )
        has_conflicting_power = bool(
            expected_power and observed_power and not (expected_power & observed_power)
        )

        # A factory code is proposed only with a product-family match plus a
        # shared technical value, or with a very strong textual match.
        can_keep_code = code_is_available and (
            (shared_specs and shared_tokens >= 2 and score >= 0.25) or score >= 0.52
        ) and not has_conflicting_voltage and not has_conflicting_power
        review = score >= 0.18 and not can_keep_code
        if can_keep_code:
            status = 'CONSERVAR CÓDIGO DE FÁBRICA'
            internal_qr = ''
        elif review:
            status = 'REVISAR COINCIDENCIA'
            internal_qr = ''
        else:
            status = 'GENERAR QR INTERNO'
            internal_qr = f"PROENERGIM-{int(product['ITEM']):03d}"

        reconciliation.append({
            'item': int(product['ITEM']),
            'almacen': product['ALMACEN'],
            'tipo_oficial': product['TIPO'],
            'producto_oficial': product['PRODUCTO'],
            'unidad': product['UM'],
            'stock_excel': int(product['UND']) if pd.notna(product['UND']) else 0,
            'foto_referencia': int(best['n']) if status != 'GENERAR QR INTERNO' else '',
            'archivo_foto': best['archivo'] if status != 'GENERAR QR INTERNO' else '',
            'codigo_fabrica_propuesto': code if can_keep_code else '',
            'tipo_codigo': best['tipo_codigo'] if can_keep_code else '',
            'marca_propuesta': best['marca'] if status != 'GENERAR QR INTERNO' else '',
            'modelo_propuesto': best['modelo'] if status != 'GENERAR QR INTERNO' else '',
            'subtipo_propuesto': best['subtipo'] if status != 'GENERAR QR INTERNO' else '',
            'atributos_foto': best['datos'] if status != 'GENERAR QR INTERNO' else '',
            'especificaciones_coincidentes': ', '.join(shared_specs),
            'conflicto_especificaciones': (
                'Voltaje no coincide' if has_conflicting_voltage
                else 'Potencia no coincide' if has_conflicting_power
                else ''
            ),
            'confianza': round(score, 3),
            'estado': status,
            'qr_interno_propuesto': internal_qr,
            'nota': (
                'Código confirmado por tipo/nombre y especificaciones.' if can_keep_code
                else 'Hay una posible foto relacionada; validar antes de asignar código.' if review
                else 'No existe evidencia fotográfica suficiente; generar QR interno.'
            ),
        })

    for row in reconciliation:
        for key, value in row.items():
            if pd.isna(value):
                row[key] = ''
    OUTPUT.write_text(
        json.dumps(reconciliation, ensure_ascii=False, indent=2, allow_nan=False),
        encoding='utf-8',
    )
    candidates = []
    for _, photo in photos[photos['n'].isin(NEW_CANDIDATE_PHOTOS)].iterrows():
        candidates.append({
            'foto': int(photo['n']),
            'producto_propuesto': photo['subtipo'],
            'categoria': photo['categoria'],
            'marca': photo['marca'],
            'modelo': photo['modelo'],
            'codigo_fabrica': '' if str(photo['codigo']).strip() == '-' else photo['codigo'],
            'datos_tecnicos': photo['datos'],
            'estado': 'PENDIENTE DE APROBACIÓN',
            'motivo': NEW_CANDIDATE_PHOTOS[int(photo['n'])],
        })
    NEW_CANDIDATES_OUTPUT.write_text(
        json.dumps(candidates, ensure_ascii=False, indent=2, allow_nan=False),
        encoding='utf-8',
    )
    print(f'Wrote {len(reconciliation)} official records to {OUTPUT}')
    print(pd.Series([row['estado'] for row in reconciliation]).value_counts().to_string())


if __name__ == '__main__':
    main()
